File: ui.apps/src/main/content/jcr_root/apps/granite/ui/components/shell/clientlibs/shell/js/collectionpage/sort.js
/*************************************************************************
* ADOBE CONFIDENTIAL
* ___________________
*
* Copyright 2016 Adobe
* All Rights Reserved.
*
* NOTICE: All information contained herein is, and remains
* the property of Adobe and its suppliers, if any. The intellectual
* and technical concepts contained herein are proprietary to Adobe
* and its suppliers and are protected by all applicable intellectual
* property laws, including trade secret and copyright laws.
* Dissemination of this information or reproduction of this material
* is strictly forbidden unless prior written permission is obtained
* from Adobe.
**************************************************************************/
(function(document, $, Coral) {
    "use strict";

    /**
     * @typedef {Object} SortConfig
     * @property {String} sortBy
     * @property {String} SortOrder
     */

    var sortContainerMap = new WeakMap();

    /**
     * Get the sortByItem whose value attribute is equal to passed value
     *
     * @param {HTMLElement} sortContainerEl container/parent of sort element
     * @param {String} value whose sortByItem needs to be retrieved
     * @returns {HTMLElement} sortByItem with provided value attribute
     */
    function getSortByItemFromValue(sortContainerEl, value) {
        var sortByEl = sortContainerEl.querySelector(".shell-collection-sortby");
        return sortByEl.items.getAll().find(function(sortByItem) {
            return sortByItem.getAttribute("value") === value;
        });
    }

    /**
     * Get the sortOrderItem whose value attribute is equal to passed value
     *
     * @param {HTMLElement} sortContainerEl container/parent of sort element
     * @param {String} value whose sortOrderItem needs to be retrieved
     * @returns {HTMLElement} sortOrderItem with provided value attribute
     */
    function getSortOrderItemFromValue(sortContainerEl, value) {
        var sortOrderEl = sortContainerEl.querySelector(".shell-collection-sortorder");
        return sortOrderEl.items.getAll().find(function(sortOrderItem) {
            return sortOrderItem.getAttribute("value") === value;
        });
    }

    /**
     * Get the selected sortByItem columnName
     *
     * @param {HTMLElement} sortContainerEl container/parent of sort element
     * @returns {Object} set of API related to shell sort Container.
     */
    function getSortContainerAPI(sortContainerEl) {
        var sortByEl = sortContainerEl.querySelector(".shell-collection-sortby");
        var sortOrderEl = sortContainerEl.querySelector(".shell-collection-sortorder");
        var sortOrderButtonEl = sortOrderEl.querySelector(".shell-collection-sortorder > button");
        /**
         * Get the selected sortByItem value
         *
         * @returns {String} value of selected sortByItem
         */
        function getSelectedSortByValue() {
            return sortByEl.selectedItem.getAttribute("value");
        }

        /**
         * Get the selected sortOrderItem value
         *
         * @returns {String} value of selected sortOrderItem
         */
        function getSelectedSortOrderValue() {
            return sortOrderEl.selectedItem.getAttribute("value");
        }

        return {
            /*
             * Get the sort values of shell sort container.
             * @returns {Config} containing sort values. keys-> sortBy, sortOrder, columnName.
             */
            getValues: function() {
                return {
                    sortBy: getSelectedSortByValue() || null,
                    sortOrder: getSelectedSortOrderValue() || null
                };
            },

            /*
             * This method sets the sort values of shell sort container.
             *
             * @param sortBy value .
             * @param sortOrder value.
             * @param silent if true only sync values and not trigger sorting
             */
            setValues: function(sortBy, sortOrder, silent) {
                var sortByItem = getSortByItemFromValue(sortContainerEl, sortBy);
                var sortOrderItem = getSortOrderItemFromValue(sortContainerEl, sortOrder);

                if (!sortBy || !sortOrder) {
                    // select none option and disable order
                    sortByItem = getSortByItemFromValue(sortContainerEl, null);
                    sortOrderItem = getSortOrderItemFromValue(sortContainerEl, "asc");
                    sortOrderButtonEl.disabled = true;
                } else if (sortByItem && sortOrderItem) {
                    sortOrderButtonEl.disabled = false;
                } else {
                    // do nothing
                    sortOrderButtonEl.disabled = false;
                    return;
                }

                if (!sortByItem.selected) {
                    sortByItem.set("selected", true, silent);
                }
                if (!sortOrderItem.selected) {
                    sortOrderItem.set("selected", true, silent);
                }
            }
        };
    }

    /**
     * Sync the sortContainer values based on collection sort values.
     *
     * @param {HTMLElement} containerEl target collection element
     * @param {HTMLElement} sortContainerEl shell sortContainer element
     */
    function syncSortContainerValues(collectionEl, sortContainerEl) {
        if (!collectionEl || !sortContainerEl) {
            return;
        }

        var sortAPI = $(collectionEl).adaptTo("foundation-collection-sort");
        var sortContainerAPI = getSortContainerAPI(sortContainerEl);

        var sortValues = sortAPI.getSortValues();

        sortContainerAPI.setValues(sortValues["sortBy"], sortValues["sortOrder"]);
    }

    /**
     * Sync the cookie value with the collection element sort values. Never sync with sortContainer values.
     *
     * @param {HTMLElement} containerEl target collection element
     * @param {HTMLElement} sortContainerEl container/parent of sort element
     */
    function syncCookie(collectionEl, sortContainerEl) {
        if (!collectionEl || !sortContainerEl) {
            return;
        }

        var consoleId = sortContainerEl.dataset.shellCollectionpageConsoleid;
        var sortAPI = $(collectionEl).adaptTo("foundation-collection-sort");
        var sortValues = sortAPI.getSortValues();

        if (!consoleId) {
            return;
        }

        var sortBykey = consoleId + "-sortName";
        var sortOrderkey = consoleId + "-sortDir";

        if (sortValues["sortBy"] && sortValues["sortOrder"]) {
            $.cookie(sortBykey, sortValues["sortBy"], { expires: 7, path: "/" });
            $.cookie(sortOrderkey, sortValues["sortOrder"], { expires: 7, path: "/" });
        } else {
            $.removeCookie(sortBykey, { path: "/" });
            $.removeCookie(sortOrderkey, { path: "/" });
        }
    }

    /**
     * returns the related sort container for collection
     *
     * @param {HTMLElement} containerEl target collection element
     * @returns {HTMLElement} sortContainerEl container/parent of sort element
     */
    function getRelatedSortContainer(collectionEl) {
        if (sortContainerMap.has(collectionEl)) {
            return sortContainerMap.get(collectionEl);
        } else {
            var matchedSortContainerEl = Array.from(document.querySelectorAll(".shell-collection-sortcontainer"))
                .find(function(sortContainerEl) {
                    return collectionEl.matches(sortContainerEl.dataset.shellCollectionTarget);
                });

            if (matchedSortContainerEl) {
                sortContainerMap.set(collectionEl, matchedSortContainerEl);
                return matchedSortContainerEl;
            }
            return null;
        }
    }

    /**
     * Removes the saved sort cookie
     *
     * @param {HTMLElement} containerEl target collection element
     * @param {HTMLElement} sortContainerEl container/parent of sort element
     */
    function clearCookie(collectionEl, sortContainerEl) {
        if (!collectionEl || !sortContainerEl) {
            return;
        }

        var consoleId = sortContainerEl.dataset.shellCollectionpageConsoleid;

        if (!consoleId) {
            return;
        }

        var sortBykey = consoleId + "-sortName";
        var sortOrderkey = consoleId + "-sortDir";

        $.removeCookie(sortBykey, { path: "/" });
        $.removeCookie(sortOrderkey, { path: "/" });
    }

    /**
     * Used to track the event performed by user.
     * This function tracks interactions with breadcrumbs
     * OMEGA Implementation
     *
     * @param {HTMLElement} sortContainerEl container/parent of sort element.
     * @param {HTMLElement} sortContainerItem sortContainerEl child element which user interacted with.
     * @param {Object} attributes optional extra information that needs to be sent
     */
    function trackEvent(sortContainerEl, sortContainerItem, attributes) {
        var trackElement = sortContainerItem.matches(".shell-collection-sortby") ? "sortBy" : "sortOrder";

        var trackData = {
            element: trackElement,
            feature: "No feature defined",
            type: "sort",
            action: "click",
            widget: {
                name: trackElement,
                type: "shell-sort"
            },
            attributes: attributes || {}
        };

        $(window).adaptTo("foundation-tracker").trackEvent(trackData);
    }


    /**
     * This will be triggered when sort component value has been changed from UI/code i.e
     * the sort container values are correct and other values needs to be updated.
     */
    $(document).on("coral-cyclebutton:change",
        ".shell-collection-sortby, .shell-collection-sortorder", function(event) {
            var self = this;
            var sortContainerEl = self.closest(".shell-collection-sortcontainer");
            var targetSelector = sortContainerEl.dataset.shellCollectionTarget;
            var collectionEl = document.querySelector(targetSelector);

            if (!targetSelector || !collectionEl) {
                return;
            }

            var collection = $(collectionEl);

            var sortContainerAPI = getSortContainerAPI(sortContainerEl);
            var sortContainerValues = sortContainerAPI.getValues();
            var sortBy = sortContainerValues["sortBy"];
            var sortOrder = sortContainerValues["sortOrder"];
            var sortAPI = collection.adaptTo("foundation-collection-sort");

            sortContainerEl.sortTimeoutId && clearTimeout(sortContainerEl.sortTimeoutId);

            // void twice execution of sorting if event occur quickly
            sortContainerEl.sortTimeoutId = setTimeout(function() {
                delete sortContainerEl.sortTimeoutId;
                sortAPI.doSort(sortBy, sortOrder) &&
                trackEvent(sortContainerEl, self, { sortBy: sortBy, sortOrder: sortOrder });
            }, 50);
        });

    /**
     * This will be triggered when sorting has been changed for collection.
     * the collection/sortAPI values are correct values and other values needs to be synced.
     */
    $(document).on("foundation-collection-sort:changed", ".foundation-collection", function() {
        var collectionEl = this;

        var sortContainerEl = getRelatedSortContainer(collectionEl);

        if (!sortContainerEl) {
            return;
        }
        // Even when sortContainer are hidden, the values/cookies should get sync.
        Coral.commons.ready(function() {
            syncSortContainerValues(collectionEl, sortContainerEl);
            syncCookie(collectionEl, sortContainerEl);
        });

        // GRANITE-31732
        sortContainerEl[collectionEl.matches(".foundation-layout-table.foundation-collection")
            ? "setAttribute" : "removeAttribute"]("hidden", "");
    });

    /**
     * This will be executed before sorting starts.
     */
    $(document).on("foundation-collection-sort:beforeChange", ".foundation-collection", function() {
        var sortAPI = $(this).adaptTo("shell-collection-sort");
        if (sortAPI) {
            sortAPI.clearCookie();
        }
    });

    /**
     * Change the sort order icon to size 'XS'
     * coral-cyclebutton does not support iconSize, this is a workaround
     */
    $(document).on("foundation-contentloaded", function(e) {
        e.target.querySelectorAll(".shell-collection-sortcontainer").forEach(function(sortContainerEl) {
            Coral.commons.ready(sortContainerEl, function() {

                if(null==$.cookie("cq-assets-files-files-sortName")){
					$.cookie("cq-assets-files-files-sortName", "name", { expires: 7, path: "/" });
                    $.cookie("cq-assets-files-files-sortDir", "asc", { expires: 7, path: "/" });
                    location.reload();

                }
                var sortOrder = sortContainerEl.querySelector(".shell-collection-sortorder");
                var sortOrderButton = sortOrder.querySelector(".shell-collection-sortorder > button");
                if (sortOrderButton && sortOrderButton.iconSize !== Coral.Icon.size.EXTRA_SMALL) {
                    sortOrderButton.iconSize = Coral.Icon.size.EXTRA_SMALL;
                }
            });

            var collectionEl = document.querySelector(sortContainerEl.dataset.shellCollectionTarget);

            if (collectionEl) {
                sortContainerMap.set(collectionEl, sortContainerEl);

                // GRANITE-31732
                sortContainerEl[collectionEl.matches(".foundation-layout-table.foundation-collection")
                    ? "setAttribute" : "removeAttribute"]("hidden", "");
            }
        });
    });


    /**
     * adaptTo implementation
     */
    $(window).adaptTo("foundation-registry").register("foundation.adapters", {
        type: "shell-collection-sort",
        selector: ".foundation-collection",
        adapter: function(collectionEl) {
            var sortContainerEl = getRelatedSortContainer(collectionEl);

            if (!sortContainerEl) {
                return null;
            }

            return {
                clearCookie: function() {
                    clearCookie(collectionEl, sortContainerEl);
                }
            };
        }
    });
})(document, Granite.$, Coral);

File: ui.apps/src/main/content/jcr_root/apps/granite/ui/components/shell/collectionpage/sort/Sort.java
/*************************************************************************
 * ADOBE CONFIDENTIAL
 * ___________________
 *
 * Copyright 2018 Adobe
 * All Rights Reserved.
 *
 * NOTICE: All information contained herein is, and remains
 * the property of Adobe and its suppliers, if any. The intellectual
 * and technical concepts contained herein are proprietary to Adobe
 * and its suppliers and are protected by all applicable intellectual
 * property laws, including trade secret and copyright laws.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Adobe.
 **************************************************************************/
package apps.granite.ui.components.shell.collectionpage.sort;

import com.adobe.granite.ui.components.AttrBuilder;
import org.apache.sling.api.resource.Resource;
import org.apache.sling.api.resource.ValueMap;
import javax.servlet.http.Cookie;
import com.adobe.granite.ui.components.htl.ComponentHelper;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Sort extends ComponentHelper {
    private final String DEFAULT_CONFIG_PATH = "granite/ui/content/shell/collectionpage/sort";
    private boolean ordered;
    private boolean override;
    private String customConfigPath;

    private ArrayList<SortByItem> sortByItems = new ArrayList<SortByItem>();
    private ArrayList<SortOrderItem> sortOrderItems = new ArrayList<SortOrderItem>();
    private static final Logger LOGGER = LoggerFactory.getLogger(Sort.class);

    @Override
    public void activate() {
        ValueMap vm  = this.getResource().getValueMap();
        ordered = vm.get("ordered", false);
        override = vm.get("override", false);
        customConfigPath = vm.get("configPath", String.class);

        Resource customResource = this.getResourceResolver().getResource(customConfigPath);
        Resource defaultResource = this.getResourceResolver().getResource(DEFAULT_CONFIG_PATH);

        addSortByItems(defaultResource.getChild("sortBy"), override);
        addSortOrderItems(defaultResource);
        if(customResource != null){
            addSortByItems(customResource, false);
        }
    }

    public ArrayList<SortByItem> getSortByItems() {
        if(ordered) {
            Collections.sort(sortByItems);
        }
        return new ArrayList<>(sortByItems);
    }

    public ArrayList<SortOrderItem> getSortOrderItems() {
        return new ArrayList<>(sortOrderItems);
    }

    public Map<String, String> getAttributes() {
        AttrBuilder attrs = getInheritedAttrs();
        populateCommonAttrs(attrs);
        return attrs.getData();
    }

    public String getSelectedSortByValue() {
        SortByItem selectedItem = null;
        for(SortByItem item : sortByItems){
            if(item != null && item.getSelected()){
                selectedItem = item;
                break;
            }
        }
        if(null == selectedItem){
            return getCookie(getConsoleId() + "-sortName");
        } else {
            return selectedItem.getValue();
        }
    }

    public String getSelectedSortOrderValue() {
        SortOrderItem selectedItem = null;
        for(SortOrderItem item : sortOrderItems){
            if(item != null && item.getSelected()){
                selectedItem = item;
                break;
            }
        }
        if(null == selectedItem){
            return getCookie(getConsoleId() + "-sortDir");
        } else {
            return selectedItem.getValue();
        }
    }

    public boolean getOrdered(){
        return ordered;
    }

    public boolean getOverride(){
        return override;
    }

    public String getLabel(){
        return "Sort by";
    }

    public String getCustomConfigPath(){
        return customConfigPath != null ? customConfigPath : "";
    }

    private void addSortByItems(Resource resource, boolean override) {
        // todo add getItemDataSource in htl componenthelper
        final Resource items = resource.getChild("items");
        if(items != null) {
            Iterator<Resource> it = items.listChildren();
            for (; it.hasNext();) {
                Resource item = it.next();
                ValueMap vm = item.getValueMap();
                String text = vm.get("text", String.class);
                String value = vm.get("value", String.class);
                String columnName = vm.get("columnName", String.class);

                if(!override || vm.get("compulsory", false)){
                    sortByItems.add(new SortByItem(text, value, vm.get("selected", "false")));
                }
            }
        }
    }

    private void addSortOrderItems(Resource resource) {
        // todo add getItemDataSource in htl componenthelper
        final Resource sortOrder = resource.getChild("sortOrder");
        Iterator<Resource> it = sortOrder.getChild("items").listChildren();
        for (; it.hasNext();) {
            Resource item = it.next();
            ValueMap vm = item.getValueMap();
            String text = vm.get("text", String.class);
            String value = vm.get("value", String.class);
            String icon = vm.get("icon", String.class);
            sortOrderItems.add(new SortOrderItem(text, value, icon));
        }
    }

    public class SortByItem implements Comparable<SortByItem> {
        private final String label;
        private final String value;
        private boolean selected;

        SortByItem(String label, String value, String columnName) {
            this.label = label;
            this.value = value;
            this.selected = false;
            if(columnName.equalsIgnoreCase("true")){
				this.selected = true;
            }
        }

        public String getLabel() {
            return label;
        }

        public String getValue() {
            return value;
        }

        public boolean getSelected() {
            return selected;
        }

        public int compareTo(SortByItem item) {
            return this.getLabel().compareTo(item.getLabel());
        }
    }

    public class SortOrderItem {
        private final String label;
        private final String value;
        private final String icon;
        private boolean selected;

        SortOrderItem(String label, String value, String icon) {
            this.label = label;
            this.value = value;
            this.icon = icon;
            this.selected = value.equals(getSelectedSortOrderValue());
        }

        public String getLabel() {
            return label;
        }

        public String getValue() {
            return value;
        }

        public boolean getSelected() {
            return selected;
        }

        public String getIcon() {
            return icon;
        }
    }

    private String getConsoleId() {
        AttrBuilder attrs = getInheritedAttrs();
        return attrs.getData().get("data-shell-collectionpage-consoleid");
    }

    private String getCookie(String key) {
        try {
            Cookie cookie = this.getRequest().getCookie(key);

            if (cookie == null) {
                return null;
            }

            return URLDecoder.decode(cookie.getValue(), "utf-8");
        } catch (UnsupportedEncodingException impossible) {
            throw new RuntimeException(impossible);
        }
    }

}
